[ADD] auth_from_http_remote_user_role

TDu opened the pull request:
A new module to assign user role defined in the http header
It is based on auth_from_http_remote_user  #7

TDu committed:
[ADD] auth_from_http_remote_user_role

guewen commented on auth_from_http_remote_user_role/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    'version': '11.0.1.0.0',
+    'category': 'Tools',
+    'website': 'https://github.com/OCA/server-auth',
+    'author': 'Thierry Ducrest, Odoo Community Association (OCA)',
```
Author should be `Camptocamp, Odoo Community Association (OCA)`

guewen commented on auth_from_http_remote_user_role/controllers/main.py:
```diff
@@ -0,0 +1,49 @@
+            roles = roles_in_header.split(self._REMOTE_USER_ROLE_SEPARATOR)
+        return roles
+
+    def _authenticate_user(self, user, cr):
```
What is the reason to use direct SQL rather than the ORM? if there is a reason, could you document it as comments? Besides, cache must be invalidated after the executes.

guewen commented on auth_from_http_remote_user_role/controllers/main.py:
```diff
@@ -0,0 +1,49 @@
+    def _authenticate_user(self, user, cr):
+        """Update roles assigned to user """
+        cr.execute('''DELETE FROM res_users_role_line
+                        WHERE user_id=%s''', [user.id])
```
I don't think we should delete the roles every time if they don't change. It increases the risk of concurrency errors and it's doing useless delete/writes. Most of the time, the roles won't change.

sbidoul replied:
Be careful with performance: changing roles implies writing to user groups which implies cache invalidation.

TDu commented:
Back at it, thank you for the reviews.
I removed all the direct sql queries and made sure that the roles where updated only when needed.
Please have another look.

asaunier commented on auth_from_http_remote_user_role/README.rst:
```diff
@@ -0,0 +1,80 @@
+Contributors
+------------
+
+* Thierry Ducrest <thierry.ducrest@camptocamp.com
```
Missing trailing ``>``

asaunier commented on auth_from_http_remote_user_role/tests/tests_user_role_change.py:
```diff
@@ -0,0 +1,33 @@
+class TestUserRoleChange(TransactionCase):
+
+    def setUp(self):
+        super(TestUserRoleChange, self).setUp()
```
Could be written ``super().setUp()``

asaunier commented on auth_from_http_remote_user_role/tests/tests_user_role_change.py:
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
```
No longer needed in Python 3

asaunier reviewed (commented):
I have spotted only very minor points. Else LGTM :)

TDu commented:
Thanks made the changes

TDu commented:
Waiting on #7 to be merged to finish this one

p-tombez commented:
@TDu When testing this module, I found a bug when you try to remove user roles: the server will return a 500 error. Here is a fix : https://github.com/TDu/server-auth/pull/1

p-tombez committed:
Fix crash when removing user role (#1)

Commit:
Rework update of roles from HTTP header

* the code was never called because of a typo on login_http_remote_user
(logging_http_remote_user)
* Updating the roles at login was not enough, as the session duration is
extensively long, so they are checked at each request
* the check is light because we store the last header in res_users, only
if it changes we compare and modify the actual roles
* the check is triggered in _authenticate as it is called for every
request and we have to run it only if the authentication was successful

simahawk commented on auth_from_http_remote_user_role/tests/tests_user_role_change.py:
```diff
@@ -0,0 +1,37 @@
+
+class TestUserRoleChange(TransactionCase):
+
+    def setUp(self):
```
any chance to see ir.http tests here? :innocent:

simahawk reviewed (commented):
Looks mostly good. Would be nice to see some more test cov.
Not sure is related to this PR but tests are failing on auth_brute_force

simahawk commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,83 @@
+            return user[0]
+        return None
+
+    def logging_http_remote_user(self, env, user):
```
@TDu @guewen I think this meth is never used since it has been renamed https://github.com/OCA/server-auth/pull/7/files#diff-00f6cb1d0bf2f276bf85d1ccf7615dfbR43

Commit:
Fix roles lost when the header is missing from requests

Use case:

* Odoo requests itself asking for a report; no SSO header present
* Direct access to odoo bypassing the proxy (for debugging only); no SSO
  header present

In these situations, the roles are lost each time such a request
hits Odoo. This is specially noticeable when a report is printed:
Odoo requests itself, as we have no header, the roles are removed from
the user, so an access error is triggered during the report generation,
but when the user browse on Odoo (through the SSO proxy), the roles are
instantly back.

Now, the role changes will be applied only if the header is
present (empty or not). It makes sense because we can consider that
if the header is absent, it comes from a trusted environment. The
reverse proxy enforces the header for the untrusted world.

grindtildeath reviewed (approved):
LGTM :+1:

TDu commented:
@guewen  :+1:  Thanks for your help and the great improvements on this.
https://github.com/OCA/server-auth/pull/9/files#diff-d8f62dd3485592027fe4a40b3db0b97dR55 -> That is quiet enlightening :zap:

simahawk commented on auth_from_http_remote_user_role/models/ir_http.py:
```diff
@@ -0,0 +1,131 @@
+            new_roles = set()
+            role_codes = cls._get_http_role_codes(roles_in_header)
+            if role_codes:
+                new_roles = set(env['res.users.role'].search(
```
small detail: I don't think you need to call set() on .ids as they should be already unique

simahawk reviewed (approved):
LG, would be nice to see more test cov :pray:
